core/query/mixingQuery.ml

```diff
@@ -0,0 +1,1551 @@
+               is defined as:
+
+               fun asList(t) server {for (x <-- t) [x]}
+            *)
+            | v ->
+              (* In order to maintain the invariant that each
+                 bound variable is unique we freshen all for-bound
+                 variables in v here.
+
+                 This is necessary in order to ensure that each
+                 instance of a table in a self-join is given a
+                 distinct alias, as the alias is generated from the
+                 name of the variable binding the table.
+
+                 We are assuming that any closure-bound variables will
+                 be eliminated anyway.
+              *)
+              (* Debug.print ("env v: "^string_of_int var^" = "^string_of_t v); *)
+              Q.freshen_for_bindings (Env.Int.empty) v
+        end
+    | Extend (ext_fields, r) ->
+      begin
+        match opt_app (xlate env) (Q.Record StringMap.empty) r with
+          | Q.Record fields ->
+            Q.Record (StringMap.fold
+                       (fun label v fields ->
+                         if StringMap.mem label fields then
+                           query_error
+                             "Error adding fields: label %s already present"
+                             label
+                         else
+                           StringMap.add label (xlate env v) fields)
+                       ext_fields
+                       fields)
+          | _ -> query_error "Error adding fields: non-record"
+      end
+    | Project (label, r) -> Q.Project (xlate env r, label)
+    | Erase (labels, r) -> Q.Erase (xlate env r, labels)
+    | Inject (label, v, _) -> Q.Variant (label, xlate env v)
+    | TAbs (_, v) -> xlate env v
+    | TApp (v, _) -> xlate env v
+
+    | XmlNode (tag, attrs, children) ->
+        (* TODO: deal with variables in XML *)
+        let children =
+          List.fold_right
+            (fun v children ->
+               let v = xlate env v in
+               List.map Q.unbox_xml (Q.unbox_list v) @ children)
+            children [] in
+        let children =
+          StringMap.fold
+            (fun name v attrs ->
+               Value.Attr (name, Q.unbox_string (xlate env v)) :: attrs)
+            attrs children
+        in
+          Q.Singleton (Q.XML (Value.Node (tag, children)))
+
+    | ApplyPure (f, ps) ->
+        reduce_artifacts (Q.Apply (xlate env f, List.map (xlate env) ps))
+    | Closure (f, _, v) ->
+      let open Lang in
+      let (_finfo, (xs, body), z_opt, _location) = Tables.find Tables.fun_defs f in
+      let z = OptionUtils.val_of z_opt in
+      (* Debug.print ("Converting evalir closure: " ^ Var.show_binder (f, _finfo) ^ " to query closure"); *)
+      (* yuck! *)
+      let env' = bind (empty_env env.policy) (z, xlate env v) in
+      Q.Closure ((xs, body), env')
+      (* (\* Debug.print("looking up query closure: "^string_of_int f); *\) *)
```

Comment:
Are these commented out lines still relevant?

Reply:
No, I'm removing this.